### ISSUE TITLE
Can IDE work with LetsEncrypt and ide.example.com URLs?

### ISSUE THREAD

**Issue by acondura** [docksal/docksal#1505]
I'm trying to use the IDE for an existing public domain and after enabling the IDE as per https://docs.docksal.io/tools/ide and then edited ~/.docksal/stacks/overrides-ide.yml to set:
- hostname: ide.${VIRTUAL_HOST}
- expose to 443
- and labels to - io.docksal.virtual-host=ide.${VIRTUAL_HOST}
      - io.docksal.virtual-port=443
      - io.docksal.cert-name=${VIRTUAL_HOST_CERT_NAME:-none}

Restarted docksal and the project and got a 502 Bad Gateway. Apparently ide-example.com is still being used somewhere.

Any idea how I can make ide.exmple.com work and use my existing LE certs?

**Comment by acondura** [comment]
I was able to change this line:
- io.docksal.virtual-host=ide-${VIRTUAL_HOST},ide-${VIRTUAL_HOST}.*
into this:
- io.docksal.virtual-host=ide.${VIRTUAL_HOST}
in file ~/.docksal/stacks/overrides-ide.yml and it worked, ide.example.com loaded with SSL and the coder interface.

I also tried updating PROJECT_ROOT/.docksal/docksal.yml to use:

version: "2.1"
services:
  web:
    labels:
      io.docksal.virtual-host: opncld.com,*.opncld.com
  ide:
    labels:
      - io.docksal.virtual-host=ide.${VIRTUAL_HOST}

but it didn't work. It's like it didn't consider it. Is it because IDE cannot be overridden in the project's yml file?

**Comment by lmakarov** [comment]
If you set `VIRTUAL_HOST=example.com`, then it should all work out of the box.
https://docs.docksal.io/core/system-vhost-proxy/#overriding-the-default-domain-name

You can also set a completely custom domains for any of the services using the `io.docksal.virtual-host` label.
https://docs.docksal.io/core/system-vhost-proxy/#custom-domains

```
version: "2.1"
services:
  web:
    labels:
      io.docksal.virtual-host:example.com,*.example.com
  ide:
    labels:
      - io.docksal.virtual-host=ide.myotherdomain.com
```

**Comment by acondura** [comment]
Well, my problem was that I already had the VIRTUAL_HOST to example.com and its subdomains, but when I followed the steps to enable IDE, it enabled ide-example.com instead of ide.example.com as expected. That's why I've overridden it in ~/.docksal/stacks/overrides-ide.yml.
I was just surprised that I could not fix it in project_root/.docksal/docksal.yml and that I had to do it in the user's .docksal. Now that I think of it, it's actually better to change that in one place (user's .docksal) than to do it in every project.

**Comment by lmakarov** [comment]
> I was just surprised that I could not fix it in project_root/.docksal/docksal.yml and that I had to do it in the user's .docksal.

Project level `docksal.yml` is loaded last and will take precedence. Maybe you did not run `fin up` / `fin project start` to update the stack configuration after editing the file? You can always check the resulting stack configuration that will be applied with `fin config show` and verify the final value for the `io.docksal.virtual-host` label on the `ide` container.

> Now that I think of it, it's actually better to change that in one place (user's .docksal) than to do it in every project.

That would be like patching Drupal's core - better to not do it that way. Also, such changes will get reset with each `fin update`.

**Comment by acondura** [comment]
Yeah, that's the first thing I did. I've just ran the following commands in sequence and I still get ide-example.com instead of ide.example.com:

1. fin update
2. set the project's docksal.yml to:

`
version: "2.1"
services:
  ide:
    labels:
      - io.docksal.virtual-host=ide.${VIRTUAL_HOST}
`

3. run `fin up`, still no change, run `fin p start`, still no change.

**Comment by acondura** [comment]
Could it be that fin is ignoring somehow the project's IDE overrides?

**Comment by acondura** [comment]
Would this have anything to do with the fact that the project's IDE settings are ignored?

![Screenshot from 2021-06-25 16-45-32](https://user-images.githubusercontent.com/1694890/123434405-1a3eb680-d5d5-11eb-9c03-17d7c65a520f.png)

**Comment by acondura** [comment]
Or maybe this:

![Screenshot from 2021-06-25 16-49-17](https://user-images.githubusercontent.com/1694890/123434599-583bda80-d5d5-11eb-9f4a-61cca35a710e.png)

**Comment by acondura** [comment]
The winning code fix was:
 ```
               # Enable VSCode (Coder) IDE
                overrides_ide_file="$(get_config_dir_dc)/stacks/overrides-ide.yml"
                if [[ -f "${overrides_ide_file}" ]] && [[ "$IDE_ENABLED" != "" ]] && [[ "$IDE_ENABLED" != "0" ]]; then
                        COMPOSE_FILE="${COMPOSE_FILE}${SEPARATOR}${overrides_ide_file}"
                fi

                # Override global VSCode (Coder) IDE settings
                project_ide_file="$(get_project_path_dc)/.docksal/docksal-ide.yml"
                if [[ -f "${project_ide_file}" ]] && [[ "$IDE_ENABLED" != "" ]] && [[ "$IDE_ENABLED" != "0" ]]; then
                        COMPOSE_FILE="${COMPOSE_FILE}${SEPARATOR}${project_ide_file}"
                fi
```
All that's left now is to update the output bit as well (e.g. Web IDE URL: http://ide-example.com)

**Comment by lmakarov** [comment]
After further checking thing, I see that we load the overrides yaml files last:

```
❯ fin config show
---------------------
COMPOSE_PROJECT_NAME_SAFE: test
COMPOSE_FILE:
/Users/leonid/.docksal/stacks/volumes-bind.yml
/Users/leonid/.docksal/stacks/stack-default.yml
/Users/leonid/Work/Projects/test/.docksal/docksal.yml
/Users/leonid/.docksal/stacks/overrides-dd-bind.yml
/Users/leonid/.docksal/stacks/overrides-ide.yml
...
```

As such, overriding the the overrides is not currently possible.

We could add support for another optional config file that is loaded last and thus can override all previous settings (similar to what you suggested above).

Alternatively (and for the time being), you can include the IDE configuration as part of the project stack:

- `fin config remove IDE_ENABLED`
- copy over the contents of `~/.docksal/stacks/overrides-ide.yml` into project's `.docksal/docksal.yml`
- adjust config as needed
- `fin project start`

You could also put this configuration into an environment specific file and active it when needed:

```
cp ~/.docksal/stacks/overrides-ide.yml .docksal/docksal-ide.yml
# adjust config as needed
fin config set DOCKSAL_ENVIRONMENT=ide

❯ fin config show
---------------------
COMPOSE_PROJECT_NAME_SAFE: test
COMPOSE_FILE:
/Users/leonid/.docksal/stacks/volumes-nfs.yml
/Users/leonid/.docksal/stacks/stack-default.yml
/Users/leonid/Work/Projects/test/.docksal/docksal.yml
/Users/leonid/Work/Projects/test/.docksal/docksal-ide.yml
ENV_FILE:
/Users/leonid/Work/Projects/test/.docksal/docksal.env
...
```

**Comment by acondura** [comment]
Interesting, very good alternatives, thank you.
Couldn't the project's .docksal/docksal.yml (and the docksal-ENV.yml ones) be loaded last since those (as far as I understand it) override the existing stacks anyway, stacks, which are "initialized" anyway by the env var in docksal.env?

**Comment by acondura** [comment]
The first alternative doesn't work, it doesn't even pick up that there is an `ide:` service in the yml file. Probably because it checks if IDE_ENABLED is 1 and if it is, then it uses the overrides-ide.yml file.

Trying the second alternative next.

**Comment by acondura** [comment]
With the first alternative I used to get 502 Bad Gateway. With the second alternative I don't get it anymore but ide.example.com loads example.com because it still loads the user's overrides-ide.yml file. So, it looks like there's no getting around the overrides-ide.yml for now and I think the reason for that is that the IDE environment is dependent on IDE_ENABLED=1 (which it should be) and if that's 1 then it uses the hard-coded path of $(get_config_dir_dc)/stacks/overrides-ide.yml.

The only way I could get it to work was if I updated these lines:
```
# Read yml for VSCode (Coder) IDE
if [[ -f "$(get_project_path_dc)/.docksal/docksal-ide.yml" ]]; then
        overrides_ide_file="$(get_project_path_dc)/.docksal/docksal-ide.yml"
else
        overrides_ide_file="$(get_config_dir_dc)/stacks/overrides-ide.yml"
fi
# Enable VSCode (Coder) IDE
if [[ -f "${overrides_ide_file}" ]] && [[ "$IDE_ENABLED" != "" ]] && [[ "$IDE_ENABLED" != "0" ]]; then
        COMPOSE_FILE="${COMPOSE_FILE}${SEPARATOR}${overrides_ide_file}"
fi
```
 I still don't get the right message displayed (e.g. Web IDE URL: http://ide-example.com) but it now works. After I fix that too, I can send a pull request if you want and if the changes make sense.

**Comment by acondura** [comment]
I've created the pull request above which takes care of:
- overriding IDE settings through the project's .docksal/docksal-ide.yml file
- overriding Xhprof settings through the project's .docksal/docksal-xhprof.yml file
